darts/timeseries.py

```diff
@@ -4233,66 +4269,138 @@ def _fill_missing_dates(
         )
 
         time_dim = xa.dims[0]
-        sorted_xa = (
+        sorted_xa = cls._sort_index(xa)
+        time_index: Union[
+            pd.Index, pd.RangeIndex, pd.DatetimeIndex
+        ] = sorted_xa.get_index(time_dim)
+
+        if isinstance(time_index, pd.DatetimeIndex):
+            has_datetime_index = True
+            observed_freqs = cls._observed_freq_datetime_index(time_index)
+        else:  # integer index (non RangeIndex)
+            has_datetime_index = False
+            observed_freqs = cls._observed_freq_integer_index(time_index)
+
+        offset_alias_info = (
+            (
+                " For more information about frequency aliases, read "
+                "https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#offset-aliases"
+            )
+            if has_datetime_index
+            else ""
+        )
+        if not len(observed_freqs) == 1:
+            raise_log(
+                ValueError(
+                    f"Could not observe an inferred frequency. An explicit frequency must be evident over a span of "
+                    f"at least 3 consecutive time stamps in the input data. {offset_alias_info}"
+                    if not len(observed_freqs)
+                    else f"Could not find a unique inferred frequency (not constant). Observed frequencies: "
+                    f"{observed_freqs}. If any of those is the actual frequency, try passing it with "
+                    f"`fill_missing_dates=True` and `freq=your_frequency`.{offset_alias_info}"
+                ),
+                logger,
+            )
+
+        freq = observed_freqs.pop()
+
+        return cls._restore_xarray_from_frequency(sorted_xa, freq)
+
+    @staticmethod
+    def _sort_index(xa: xr.DataArray) -> xr.DataArray:
```

Comment:
Two points related to this method:
* Could we maybe call it from the constructor as well? I think there we have such a case of "sort-and-copy or copy", where we could reuse this method.
* Do we always need the copy? Looking the places from where this method is called, I think some of them may not need the `.copy()` and we could avoid copying in a few places where it's not needed. I haven't checked in details but I think the copying might be needed only in the constructor (as all factory methods end up calling the constructor, if it's done there we know that the `TimeSeries` will carry its own copy anyway and we can avoid copying earlier in the factory methods). I would recommend maybe having a `copy` boolean argument here. Given how often TimeSeries are built everywhere, avoiding these copies as much as possible could make a difference.

Reply:
👍 achieved approx. a 2% increase in time series creation time with larger series.

Reply:
Nice!